packages/block-editor/src/hooks/custom-fields.js

```diff
@@ -125,6 +163,30 @@ const withInspectorControl = createHigherOrderComponent( ( BlockEdit ) => {
 	};
 }, 'withInspectorControl' );
 
+const createEditFunctionWithPatternSource = () =>
+	createHigherOrderComponent(
+		( BlockEdit ) =>
+			( { attributes, ...props } ) => {
+				const registry = useRegistry();
+				const sourceAttributes =
+					registry._selectAttributes?.( {
+						clientId: props.clientId,
+						name: props.name,
+						attributes,
+					} ) ?? attributes;
+
+				return (
+					<BlockEdit { ...props } attributes={ sourceAttributes } />
+				);
+			}
+	);
+
+function shimAttributeSource( settings ) {
+	settings.edit = createEditFunctionWithPatternSource()( settings.edit );
```

Comment:
What's the role of this HOC?

There is a sub-registry injected in the `core/block` edit implementation that should play a similar role. It essentially should allow redefining a function that fetches the attributes for the block.

Reply:
Yeah, it was simply the minimum effort for me to make the POC work at the time. We should revisit the "reading" side of the story.